crates/clvm-traits/src/pretty_printer.rs

```diff
@@ -0,0 +1,121 @@
+use num_bigint::BigInt;
+
+use crate::{ClvmEncoder, ToClvm};
+
+const PRINTABLE: &str = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ!\"#$%&'()*+,-./:;<=>?@[]^_\\`{|}~ \t\r\n";
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct PrettyPrinter {
+    pub max_int_bytes: usize,
+    pub atom_strings: bool,
+}
+
+impl Default for PrettyPrinter {
+    fn default() -> Self {
+        Self {
+            max_int_bytes: 2,
+            atom_strings: true,
+        }
+    }
+}
+
+pub trait PrettyPrint {
+    fn pretty_print(&self) -> String;
+}
+
+impl<T> PrettyPrint for T
+where
+    T: ToClvm<Pretty>,
+{
+    fn pretty_print(&self) -> String {
+        self.to_clvm(&mut PrettyPrinter::default()).unwrap().0
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct Pretty(pub String);
+
+impl ClvmEncoder for PrettyPrinter {
+    type Node = Pretty;
+
+    fn encode_atom(&mut self, bytes: &[u8]) -> Result<Self::Node, crate::ToClvmError> {
+        if bytes.is_empty() {
+            return Ok(Pretty("()".to_string()));
+        }
+
+        if bytes.len() <= self.max_int_bytes {
+            let int = BigInt::from_signed_bytes_be(bytes);
+
+            if int.to_signed_bytes_be() != bytes {
+                return Ok(Pretty(format!("0x{}", hex::encode(bytes))));
+            }
+
+            return Ok(Pretty(BigInt::from_signed_bytes_be(bytes).to_string()));
+        }
+
+        if !self.atom_strings {
+            return Ok(Pretty(format!("0x{}", hex::encode(bytes))));
+        }
+
+        let text = String::from_utf8_lossy(bytes);
```

Comment:
it seems a bit scary to interpret the string as utf8 before ensuring all characters are printable

Reply:
The replacement character for invalid UTF-8 isn't in the printable list